README
----------------------------------
* Linux ISG :: lISG-0.11.3-alpha *
----------------------------------

This is the first public release. Nevertheless "Linux ISG" is working
for half a year under a heavy load (> 250K pps, > 5K sessions, all
sessions are bandwidth policed) without any stability or performance
issues. Also there are no (or little) complains regarding policing
precision (minimum policing rate was 200 Kbit/s).

----------------
* Installation *
----------------
0) Unpack distribution
1) cd kernel
2) Run ./configure

    *** NOTE! *** Version shown in `iptables -V' must be equal to version
    of your iptables sources.
    *** NOTE! *** If you're using iptables version < 1.4.0, you must replace
    libipt_ISG.c file with its modified version from "old_iptables" directory

3) Run make && make install
4) Run depmod -a && modprobe ipt_ISG
5) cd ..
6) Check that you have installed this Perl modules:
    Net::Radius::Packet
    Net::Radius::Dictionary
    Net::IPAddress

    *** NOTE! *** IPTables::libiptc and Time::HiRes are not needed now!

7) Copy the contents of ISG directory to your system, f.e. /usr/local/ISG
8) Change __INSTALL_DIR__/etc/config.pl to suit your needs
9) Fire up daemon ISGd.pl: __INSTALL_DIR__/bin/ISGd.pl
10) Use __INSTALL_DIR__/bin/ISG.pl to manage your sessions (try --help)

---------
* Usage *
---------
First of all you need to decide, which traffic will be managed by lISG. It
may be one subnet or multiple subnets, no matter. Imagine, that you
want to policy traffic for networks 10.254.0.0/16 and 192.168.99.0/24,
then you must do:

# iptables -A FORWARD -s 10.254.0.0/16 -j ISG --session-init
# iptables -A FORWARD -d 10.254.0.0/16 -j ISG
# iptables -A FORWARD -s 192.168.99.0/24 -j ISG --session-init
# iptables -A FORWARD -d 192.168.99.0/24 -j ISG

Rules labeled with "--session-init" parameter is so called
"initialization" rules. When packet hits this rule, ipt_ISG kernel module
tries to detect if this packet is belongs to "approved" session. If
not, ipt_ISG drop packet and send NETLINK message to ISGd.pl. Daemon
contacts RADIUS server to authorize this packet using its src IP address
as User-Name attribute. If RADIUS accepting this request, session turn
its state to "approved", daemon applies bandwidth policy to the session
and starts accounting. Otherwise session remains in "unapproved" state
for approve_retry_interval time (see ipt_ISG module parameters).

Remember that rules with "--session-init" parameter is using for session
initialization and router's inbound (FROM user) traffic policing and
accouting. Rules without this parameter is using just for outbound (TO
user) traffic policing and accounting (but only if session for this dst
IP is "approved", otherwise packets will be dropped).

You can also do something like this:

# iptables -I FORWARD -s 192.168.99.0/24 -d 213.180.204.0/24 -j ISG --session-init
# iptables -I FORWARD -d 192.168.99.0/24 -s 213.180.204.0/24 -j ISG

Here we're policing and accounting traffic just with one of Yandex
subnets :) and not touching any other traffic.

You can use ISG.pl program to manage your sessions. Try 'ISG.pl help'
for parameters.

--------------------
* Services support *
--------------------
Each session can have services applied to it. Let's take an example.
User 10.11.22.33 must have access to:
    1) Private subnets at speed of 50 Mbit/s
    2) Some specific server (10.55.11.16) with no speed limit
    3) All other addresses not going into 1 and 2 (Internet actually)
       at speed 1 Mbit/s

Firstly we must define classes in etc/tc.conf file:

    FIRST	192.168.0.0/16
    FIRST	172.16.0.0/12
    CLASS_A	10.0.0.0/8

    SECOND	10.55.11.16/32

    THRID	0.0.0.0/0

10.0.0.0/8 prefix must be named "FIRST" too. "CLASS_A" is just for
demonstration that we can apply more than one class per service. Ah,
you also may use lower case letters, I don't know why I'm using caps. :)

Next define three service descriptions (see etc/config.pl for more
parameters and some comments):

    $cfg{srv}{"PRIVATE"}{download_rate}      = "50000000;75000000";
    $cfg{srv}{"PRIVATE"}{upload_rate}        = "50000000;75000000";
    $cfg{srv}{"PRIVATE"}{traffic_classes}    = [ "CLASS_A", "FIRST" ];

    $cfg{srv}{"SSERVER"}{download_rate}      = "0;0";
    $cfg{srv}{"SSERVER"}{upload_rate}        = "0;0";
    $cfg{srv}{"SSERVER"}{traffic_classes}    = [ "SECOND" ];

    $cfg{srv}{"INET-1MBIT"}{download_rate}   = "1000000;1500000";
    $cfg{srv}{"INET-1MBIT"}{upload_rate}     = "1000000;1500000";
    $cfg{srv}{"INET-1MBIT"}{traffic_classes} = [ "THIRD" ];

User 10.11.22.33 profile at RADIUS server must contain attributes:

    Cisco-Account-Info  = APRIVATE
    Cisco-Account-Info += ASSERVER
    Cisco-Account-Info += AINET-1MBIT

Note that if you exclude last attribute (1 Mbit Inet service) then user
10.11.22.33 will not have Internet access. If you exclude PRIVATE and
SSERVER services, user will not have access to local resources but ONLY
access to 1 Mbit Internet (INET-1MBIT). User will NOT have access to
ALL resources at speed of 1 Mbit/s, only 1 Mbit to Internet. This is very
important. Also see last comment in etc/tc.conf.example about 0.0.0.0/0
meaning.

You can also define additional service, INET-5MBIT for example:

    $cfg{srv}{"INET-5MBIT"}{download_rate}   = "5000000;7500000";
    $cfg{srv}{"INET-5MBIT"}{upload_rate}     = "5000000;7500000";
    $cfg{srv}{"INET-5MBIT"}{traffic_classes} = [ "THIRD" ];

and add it to profile:

    Cisco-Account-Info  = APRIVATE
    Cisco-Account-Info += ASSERVER
    Cisco-Account-Info += AINET-1MBIT
    Cisco-Account-Info += NINET-5MBIT

Please note "N" letter before service name. This means that service will
not be active by default and can be manually activated when needed using
CoA (CoA interface for services is not impemented in 0.11).

----------------------------------
* Traffic classification support *
----------------------------------
Use etc/tc.conf file for subnet declarations. See tc.conf.example for
reference. tc.conf file is re-reading automatically, see tc_check_interval
in etc/config.pl.

You can change ipt_ISG kernel module parameter "nehash_key_len". It is 20
by default. This parameter defines hash key length. Let's consider that
this parameter is set to 16 (for simplicity) and we're adding this prefixes:

    10.5.5.0/24
    10.5.6.0/24
    192.168.0.0/16

First of all a hash with 2^16 = 65535 buckets will be created. Then:

    10.5.5.0/24		go to bucket 10.5.0.0	(10.5.5.0 & 255.255.0.0)
    10.5.99.0/24	go to bucket 10.5.0.0
    192.168.0.0/16	go to bucket 192.168.0.0

So there are already 2 elements in 10.5.0.0 bucket. For best performance
we must try to minimize number of elements in each bucket. So we can
increase nehash_key_len, but this leads to more memory usage (f.e. when
nehash_key_len=24 and adding some prefix with /8 mask we must create
65535 duplicates of this /8 prefix to fill the hash).

nehash_key_len=20 (by default) is the best trade-off between memory usage
and performance for rather sparse prefix lists.

The idea of hash is derived from Robert Olsson's fib_hash2.

--------------------------------------------------
* Supported Attributes for Access-Accept packets *
--------------------------------------------------

Attribute:	Acct-Interim-Interval
Description:	Standard attribute (as per RFC)
Example:	Acct-Interim-Interval = 180

Attribute:	Session-Timeout
Description:	Standard attribute (as per RFC)
Example:	Session-Timeout = 86400

Attribute:	Idle-Timeout
Description:	Standard attribute (as per RFC)
Example:	Session-Timeout = 1800

Attribute:	Class = "<in_rate>/<out_rate>"
Description:	Traffic rate policing for session. in_rate and out_rate is integer
		given in Kbit/s. This attribute has no effect if there is at least
		one service applied for session.
Example:	Class = "5000/2500" (5 Mbit/s for download, 2.5 for upload)

Attribute:	Framed-IP-Address = <IP-address>
Description:	See 1-to-1 NAT trick below
Example:	Framed-IP-Address = 62.117.85.254

Attribute:	Cisco-Account-Info = A<service name>
Description:	Service name for session (as defined in config.pl). Service will
		be applied for session with on status. Attribute can have multiple
		values, so one session can have as many services as you want.
		Please note, that you can't apply two or more "on" services with
		overlapping classes. There is no such restriction for "N"-version
		of this attribute (see below).
Example:	Cisco-Account-Info = AINTERNET
		Cisco-Account-Info = ALOCAL

Attribute:	Cisco-Account-Info = N<service name>
Description:	The same as A<service name> but service will be created with off
		status, so the traffic that falls into this service classes will
		be dropped. Service can be enabled later using CoA interface (but
		CoA interface for services is not impemented in 0.11).
Example:	Cisco-Account-Info = NLOCAL

-----------------------
* CoA and PoD support *
----------------------- 
lISG is supporting CoA and PoD RADIUS messages processing. First of all
you must set CoA-related settings in config.pl. You can use any of this
attributes to identify session:

    * User-Name (actually an IP-address of user)
    * Acct-Session-Id
    * NAS-Port

In "Change of Authorization" messages you can currently change only
this attributes of session:

    * Class (the attribute is using in lISG for rate policing)

-------------------
* Tips and Tricks *
-------------------
To enable reliable communications between userspace and kernel, do:

    echo "1048576" > /proc/sys/net/core/rmem_max

Otherwise there will be some problems printing long session lists.

* * * * *

You can enable static 1-to-1 NAT for selected sessions by uncommenting

    $cfg{static_nat} = 1;

in config.pl and passing Framed-IP-Address attribute in Access-Accept,
but first you need to reference tree-like structures automatically
created using IPTables::SubnetSkeleton, f.e.:

    iptables -t nat -I PREROUTING  -j stat_nat_dst_skeleton
    iptables -t nat -I POSTROUTING -j stat_nat_src_skeleton

Also, blackhole route for Framed-IP-Address is automagically created
in such scheme for easy route redistribution using Quagga:

router bgp 65505
    redistribute kernel

Note that Quagga needs this patch
    http://www.gossamer-threads.com/lists/quagga/dev/20413
for proper handling of kernel blackhole routes.

--------------
* Questions? *
--------------
E-mail me: sysoleg@yandex.ru


CHANGES
-------------------------------
lISG-0.11.3-alpha :: 2010-07-08
-------------------------------
* Revert fix from 0.11.2-alpha. The reason of error message was
  different, so this fix is pointless.
* Fix the race condition between sending events to main isg and
  reload_tc processes

-------------------------------
lISG-0.11.2-alpha :: 2010-07-07
-------------------------------
* Wrong error message during 'ISG.pl clear' command execution fixed

-------------------------------
lISG-0.11.1-alpha :: 2010-07-07
-------------------------------
* Framed-IP-Address attribute was ignored (regression in 0.11), fixed

-----------------------------
lISG-0.11-alpha :: 2010-07-06
-----------------------------
* IPv4 subnets classification support (refer to README file)

* Services support (refer to README file)

* ISG.pl show_all parameter removed, all sessions are in main list now

* ISG.pl session list output modified

* Use 127.0.0.1 NAS-IP-Address when there is no way to resolve local
  system IP-address

* AVPair's client-mac-address attribute added in RADIUS-requests

* Cisco-Control-Info attribute for <I>nput and <O>utput octets added in
  accounting requests (alongside with Acct-(In|Out)put-Octets/Gigawords)

* $cfg{default_rate} removed, please use explicit "Class" attribute or
  apply service

* All sessions has an unique ID now (even unapproved)

-----------------------------
lISG-0.10-alpha :: 2010-06-09
-----------------------------
* Support for "Acct-Interim-Interval", "Session-Timeout" and "Idle-Timeout"
  attributes in Access-Accept packets

* $cfg{alive_interval} was renamed to $cfg{session_alive_interval}

* $cfg{unlim_alive_interval} was removed permanently, please use proper
  "Acct-Interim-Interval" attribute in RADIUS Access-Accept packets

* Module parameters "session_max_duration" and "session_timeout" was removed,
  please use $cfg{session_max_duration} and $cfg{session_idle_timeout} from
  config.pl

* Don't install 1-to-1 NAT entry when Framed-IP-Address equals to User-name
  (thanks to Abram)

* Do the full session stop and start with re-authentication in case of
  Session-Timeout (previously it looks like accounting counters reset and
  only RADIUS Stop/Start)

* Don't include "Password" attribute in RADIUS accounting requests

* Minor bug fixes and code cleanups

----------------------------
lISG-0.9-alpha :: 2010-04-25
----------------------------
Userspace:
* Using "dev lo" routes for dropping traffic considered incorrect in Linux
  kernel, use blackhole routes now (see README, 1-to-1 NAT section)

Kernel module:
* Potential (impossible?) null-pointer dereference fix
* 64-bit platforms support
* Minor compile warning fix

----------------------------
lISG-0.8-alpha :: 2010-01-31
----------------------------
Userspace:
* Multiple RADIUS servers support (watch config.pl for new declaration)
* $cfg{coa_secret} is back into config
* $cfg{default_speed} is renamed to $cfg{default_rate}
* Child processes auto restart
* Minor bug fixes

Kernel module:
* Zero session counters when its moved from unapproved to approved state

----------------------------
lISG-0.7-alpha :: 2010-01-11
----------------------------
Userspace:
* $cfg{coa_secret} is removed, please use $cfg{radius_secret}
* Change the meaning of $cfg{coa_server}, see config.pl
* Authenticator check for incoming CoA requests according to RFC 2866

----------------------------
lISG-0.6-alpha :: 2010-01-10
----------------------------
Userspace:
* CoA and PoD support (RFC 3576, RFC 5176), see README for reference
* New config.pl option $cfg{nas_identifier} for NAS-Identifier RADIUS attribute
  initialization. If not set, NAS-Identifier will be equal to NAS-IP-Address.
* Minor RADIUS-request attributes rearrange
* `--session-init name' turned into just `--session-init'. Traffic classification
  (f.e. local traffic and internet traffic) will be done in different way in
  future releases.
* All Virtual ports are now referred as "Virtual" type, not "Async"
* Static NAT must be enabled explicitly in config.pl ($cfg{static_nat}, see README)
* Authenticator field in Accounting requests is now calculated according to RFC 2866

Kernel module:
* Using jhash_1word function instead of jhash2 for maximum performance
* Functionality for find session by Session-Id and session IP-address (along with
  Port-Id) was added

* Some minor bug fixes for userspace and kernel module

----------------------------
lISG-0.5-alpha :: 2009-12-28
----------------------------
* Fixed ISGd.pl initialization logic (what to do before fork() and what after this)
* Added NAS-Identifier attribute to all RADIUS packets (equals to NAS-IP-Address)
* Fixed ISG.pl behaviour when there are no sessions. Now it returns nothing.

----------------------------
lISG-0.4-alpha :: 2009-12-25
----------------------------
* lISG doesn't depend on IPTables::libiptc and Time::HiRes modules now
* "/sbin/ip" replaced to "ip" (without command path) in route addition/deletion
* Some minor bug fixes

----------------------------
lISG-0.3-alpha :: 2009-12-24
----------------------------
* This file was started
* Very nice configure script added (modified version of script from ipt_NETFLOW distribution)
* Support for >= 2.6.24 kernels (module compiles cleanly, but not tested)
* Support for iptables > 1.4.2 in libipt_ISG.c, but anyway due to IPTables::libiptc module
  restrictions you have to use iptables 1.4.2. Will be fixed in 0.4-alpha.


ISG/lib/ISG.pm
package ISG;

use strict;
use warnings;

use FindBin '$Bin';
require $Bin . "/../etc/config.pl";
our %cfg;

use POSIX;
use Socket;
use IO::Select;
use IO::Socket::INET;
use Net::Radius::Packet;
use Net::Radius::Dictionary;
use Digest::MD5;

require Exporter;
use vars qw(@ISA @EXPORT);
@ISA = qw(Exporter);

@EXPORT = qw(prepare_netlink_socket netlink_read add_socket_for_select isg_send_event isg_parse_event);

use constant AF_NETLINK 	=> 0x10;
use constant ISG_NETLINK_MAIN 	=> 31;
use constant NL_HDR_LEN 	=> 16;
use constant NLMSG_ALIGNTO	=> 4;
use constant NLMSG_DONE		=> 0x3;
use constant NLM_F_MULTI	=> 0x2;
use constant IN_EVENT_MSG_LEN 	=> 140;

use constant EVENT_LISTENER_REG  => 0x01;
use constant EVENT_SESS_APPROVE  => 0x04;
use constant EVENT_SESS_CHANGE   => 0x05;
use constant EVENT_SESS_CLEAR    => 0x09;
use constant EVENT_SESS_GETLIST  => 0x10;
use constant EVENT_SESS_GETCOUNT => 0x12;

use constant EVENT_SESS_CREATE    => 0x03;
use constant EVENT_SESS_START     => 0x06;
use constant EVENT_SESS_UPDATE    => 0x07;
use constant EVENT_SESS_STOP      => 0x08;
use constant EVENT_SESS_INFO      => 0x11;
use constant EVENT_SESS_COUNT     => 0x13;
use constant EVENT_NE_ADD_QUEUE   => 0x14;
use constant EVENT_NE_SWEEP_QUEUE => 0x15;
use constant EVENT_NE_COMMIT      => 0x16;
use constant EVENT_SERV_APPLY     => 0x17;
use constant EVENT_SDESC_ADD      => 0x18;
use constant EVENT_SDESC_SWEEP_TC => 0x19;
use constant EVENT_SERV_GETLIST   => 0x20;

use constant EVENT_KERNEL_ACK     => 0x98;
use constant EVENT_KERNEL_NACK    => 0x99;

use constant IS_APPROVED_SESSION  => (1 << 0);
use constant IS_SERVICE           => (1 << 1);
use constant SERVICE_STATUS_ON    => (1 << 2);
use constant SERVICE_ONLINE       => (1 << 3);
use constant SERVICE_NO_ACCT      => (1 << 4);
use constant IS_DYING             => (1 << 5);

use constant FLAG_OP_SET   => 0x01;
use constant FLAG_OP_UNSET => 0x02;

sub get_conf {
    return %cfg;
}

sub NLMSG_ALIGN {
    my $len = shift;
    return ( (($len)+NLMSG_ALIGNTO-1) & ~(NLMSG_ALIGNTO-1) );
}

sub NLMSG_LENGTH {
    my $len = shift;
    return (($len) + NLMSG_ALIGN(NL_HDR_LEN));
}

sub NLMSG_SPACE {
    my $len = shift;
    return NLMSG_ALIGN(NLMSG_LENGTH($len));
}

sub pack_sockaddr_nl {
    my ($pid, $group_mask) = @_;
    return pack("S2iI", AF_NETLINK, 0, $pid, $group_mask);
}

sub unpack_sockaddr_nl {
    return unpack("S2iI", shift);
}

sub pack_nlmsg {
    my ($payload, $pid) = @_;

    my $payload_len = length($payload);
    return pack("IS2I2", NLMSG_SPACE(length($payload)), 0, 0, 0, $pid) . $payload;
}

sub unpack_nlmsghdr {
    return unpack("IS2I2", shift);
}

sub long2ip {
    return inet_ntoa(pack("N", shift));
}

sub ip2long {
    return unpack("N", inet_aton(shift));
}

sub ntohl {
    return unpack("L", pack("N", shift));
}

sub load_radius_dictionary {
    my $dict_dir = shift;

    my $rad_dict = new Net::Radius::Dictionary($dict_dir);
    $rad_dict->readfile($dict_dir . ".cisco");

    return $rad_dict;
}

sub format_mac {
    my ($mac, $step) = @_;
    my @pts;

    for (my $i = 0; $i <= 8 && $step; $i += $step) {
	push(@pts, substr($mac, $i, $step));
    }

    return join(".", @pts);
}

sub hex_session_id_to_llu {
    my $session_id = shift;

    if ($session_id !~ m/^[A-F0-9]{16}$/i) {
	return 0;
    }

    return pack("I", hex(substr($session_id, 8, 16))) . pack("I", hex(substr($session_id, 0, 8)));
}

sub send_coa_request {
    my ($rad_dict, $code, $coa_secret, $coa_port, $avs) = @_;
    my $rep_data;

    my $req = new Net::Radius::Packet $rad_dict;

    $req->set_code($code);
    $req->set_attr("NAS-IP-Address" => &isg_get_nas_ip());

    foreach my $av (@{$avs}) {
	foreach my $key (keys %{$av}) {
	    $req->set_attr($key => $av->{$key});
	}
    }

    $req->set_identifier(1);
    $req->set_authenticator("\x0" x 16);
    $req->set_authenticator(Digest::MD5::md5($req->pack . $coa_secret));

    my $sock = IO::Socket::INET->new( PeerPort => $coa_port,
				      PeerAddr => "localhost",
				      Proto    => "udp"
				    ) or return 1;

    $sock->send($req->pack);

    my @ready = IO::Select->new($sock)->can_read(5);
    if (@ready) {
	if ($sock->recv($rep_data, 1500)) {
	    return new Net::Radius::Packet $rad_dict, $rep_data;
	}
    }

    return 1;
}

sub init_event_fields {
    my $pars;

    $pars->{'type'}			= 0;

    $pars->{'session_id'}		= "";
    $pars->{'ipaddr'}			= 0;
    $pars->{'nat_ipaddr'}		= 0;
    $pars->{'port_number'}		= 0;
    $pars->{'flags'}			= 0;
    $pars->{'export_interval'}		= $cfg{session_alive_interval};
    $pars->{'idle_timeout'}		= $cfg{session_idle_timeout};
    $pars->{'max_duration'}		= $cfg{session_max_duration};
    $pars->{'in_rate'}			= 0;
    $pars->{'in_burst'}			= 0;
    $pars->{'out_rate'}			= 0;
    $pars->{'out_burst'}		= 0;
    $pars->{'service_name'}		= "";
    $pars->{'flags_op'}			= 0;

    $pars->{'nehash_pfx'}		= 0;
    $pars->{'nehash_mask'}		= 0;
    $pars->{'nehash_tc_name'}		= 0;

    return $pars;
}

sub pack_event {
    my $pars = shift;
    my $trash;

    if ($pars->{'type'} == ISG::EVENT_SDESC_ADD) {

	return pack("I a32 a32 a24",
	    $pars->{'type'},
	    $pars->{'nehash_tc_name'},
	    $pars->{'service_name'}
	);

    } elsif ($pars->{'type'} == ISG::EVENT_NE_ADD_QUEUE) {

	return pack("I N2 a32 a48",
	    $pars->{'type'},
	    $pars->{'nehash_pfx'},
	    $pars->{'nehash_mask'},
	    $pars->{'nehash_tc_name'}
	);

    } else {

	return pack("I a8 N2 H12 v I8 a32 C",
	    $pars->{'type'},
	    $pars->{'session_id'},
	    $pars->{'ipaddr'},
	    $pars->{'nat_ipaddr'},
	    0, # MAC-Address is read-only
	    $pars->{'flags'},
	    $pars->{'port_number'},
	    $pars->{'export_interval'},
	    $pars->{'idle_timeout'},
	    $pars->{'max_duration'},
	    $pars->{'in_rate'},
	    $pars->{'in_burst'},
	    $pars->{'out_rate'},
	    $pars->{'out_burst'},
	    $pars->{'service_name'},
	    $pars->{'flags_op'}
	);
    }
}

sub unpack_event {
    my $pars;

    my $trash;
    my ($in_packets_lo, $in_packets_hi, $out_packets_lo, $out_packets_hi);
    my ($in_bytes_lo, $in_bytes_hi, $out_bytes_lo, $out_bytes_hi);
    my ($session_id_hi, $session_id_lo);
    my ($p_session_id_hi, $p_session_id_lo);

    (
	$pars->{'type'},
	$session_id_hi,
	$session_id_lo,
	$pars->{'ipaddr'},
	$pars->{'nat_ipaddr'},
	$pars->{'macaddr'},
	$pars->{'flags'},
	$pars->{'port_number'},
	$pars->{'export_interval'},
	$pars->{'idle_timeout'},
	$pars->{'max_duration'},
	$pars->{'in_rate'},
	$pars->{'in_burst'},
	$pars->{'out_rate'},
	$pars->{'out_burst'},
	$pars->{'duration'},
	$trash, # Padding
	$in_packets_hi,
	$in_packets_lo,
	$in_bytes_hi,
	$in_bytes_lo,
	$out_packets_hi,
	$out_packets_lo,
	$out_bytes_hi,
	$out_bytes_lo,
	$p_session_id_hi,
	$p_session_id_lo,
	$pars->{'service_name'}
    ) = unpack("I I2 N2 H12 v I8 I i I10 a32", shift);

    $pars->{'service_name'} =~ s/\000//g;
    if (!length($pars->{'service_name'})) {
	undef($pars->{'service_name'});
    }

    if ($pars->{'macaddr'} eq "000000000000") {
	undef($pars->{'macaddr'});
    }

    $pars->{'session_id'} = sprintf("%08X%08X", $session_id_lo, $session_id_hi);

    if ($p_session_id_lo || $p_session_id_hi) {
	$pars->{'parent_session_id'} = sprintf("%08X%08X", $p_session_id_lo, $p_session_id_hi);
    }

    $pars->{'in_packets'} = ($in_packets_lo * 2**32) + $in_packets_hi;
    $pars->{'out_packets'} = ($out_packets_lo * 2**32) + $out_packets_hi;
    $pars->{'in_bytes'} = ($in_bytes_lo * 2**32) + $in_bytes_hi;
    $pars->{'out_bytes'} = ($out_bytes_lo * 2**32) + $out_bytes_hi;

    return $pars;
}

sub prepare_netlink_socket {
    my $sk;

    if (!socket($sk, AF_NETLINK, SOCK_RAW, ISG_NETLINK_MAIN)) {
	goto err;
    }

    if (!bind($sk, pack_sockaddr_nl(0, 0))) {
	goto err;
    }

    setsockopt($sk, SOL_SOCKET, SO_RCVBUF, pack("L", 2097152));

    return $sk;
err:
    return -1;
}

sub netlink_read {
    my ($sk, $buffer, $size, $timeout) = @_;
    my $nread;

    eval {
        local $SIG{ALRM} = sub { die "alarm\n" };
        alarm($timeout);
        $nread = sysread($sk, $$buffer, $size);
        alarm(0);
    };

    if ($@) {
        return undef;
    } else {
        return $nread;
    }
}

sub isg_send_event {
    my ($sk, $pars, $reply) = @_;

    my $data;
    my $ev = init_event_fields();

    foreach my $key (keys %{$pars}) {
	if (defined($pars->{"$key"})) {
	    $ev->{"$key"} = $pars->{"$key"};
	}
    }

    if (!send($sk, pack_nlmsg(pack_event($ev), $$), 0, pack_sockaddr_nl(0, 0))) {
	return -1;
    }

    if (defined($reply)) {
	if (!netlink_read($sk, \$data, 16384, 10)) {
	    return -1;
	} else {
	    $ev = isg_parse_event($data);
	    foreach my $key (keys %{$ev}) {
		$reply->{"$key"} = $ev->{"$key"};
	    }
	}
    }

    return 1;
}

sub isg_parse_event {
    my $data = shift;

    my @nlhdr = unpack_nlmsghdr($data);
    my $pars = unpack_event(substr($data, NL_HDR_LEN));

    $pars->{'nlhdr_len'}   = $nlhdr[0];
    $pars->{'nlhdr_type'}  = $nlhdr[1];
    $pars->{'nlhdr_flags'} = $nlhdr[2];
    $pars->{'nlhdr_seq'}   = $nlhdr[3];
    $pars->{'nlhdr_pid'}   = $nlhdr[4];

    return $pars;
}

sub isg_get_nas_ip {
    my @uname = uname();
    my $name = gethostbyname($uname[1]);

    if ($name) {
        return inet_ntoa($name);
    }

    return;
}

sub dumper {
    my $var = shift;

    use Data::Dumper;
    print Dumper($var);
}

1;
